module/zfs/vdev_mirror.c

```diff
@@ -794,7 +802,54 @@ vdev_mirror_io_done(zio_t *zio)
 		return;
 	}
 
-	/* XXPOLICY */
+	/*
+	 * If we're scrubbing but don't have a BP available (because this
+	 * vdev is under a raidz or draid vdev) then the best we can do is
+	 * compare all of the copies read.  If they're not identical then
+	 * return a checksum error and the most likely correct data.  The
+	 * raidz code will issue a repair I/O if possible.
+	 */
+	if (zio->io_flags & ZIO_FLAG_SCRUB && zio->io_bp == NULL) {
+		abd_t *last_good_abd;
+
+		ASSERT(zio->io_vd->vdev_ops == &vdev_replacing_ops ||
+		    zio->io_vd->vdev_ops == &vdev_spare_ops);
+
+		if (good_copies > 1) {
+			last_good_abd = mm->mm_child[last_good_copy].mc_abd;
+			abd_t *best_abd = NULL;
+
+			for (c = 0; c < last_good_copy; c++) {
+				mc = &mm->mm_child[c];
+
+				if (mc->mc_error || !mc->mc_tried)
+					continue;
+
+				if (abd_cmp(mc->mc_abd, last_good_abd) != 0)
+					zio->io_error = SET_ERROR(ECKSUM);
+
+				if (best_abd == NULL &&
+				    mc->mc_vd->vdev_ops ==
+				    &vdev_draid_spare_ops) {
+					best_abd = mc->mc_abd;
+				}
+			}
+
+			abd_copy(zio->io_abd, best_abd ? best_abd :
+			    last_good_abd, zio->io_size);
+
+		} else if (good_copies == 1) {
+			last_good_abd = mm->mm_child[last_good_copy].mc_abd;
+			abd_copy(zio->io_abd, last_good_abd, zio->io_size);
+		}
+
+		for (c = 0; c < mm->mm_children; c++) {
+			mc = &mm->mm_child[c];
+			abd_free(mc->mc_abd);
+			mc->mc_abd = NULL;
+		}
```

Comment:
Isn't the mc_abd just a pointer to the io_abd? Shouldn't that abd get freed by the zio code rather than this code here?

Reply:
Actually `vdev_mirror_io_start()` allocates a new abd for this child I/O so we need to make sure to clean it up.  For a scrub with an `->io_bp` this has always happened in `vdev_mirror_scrub_done()`, but since this child IO is using `vdev_mirror_child_done()` we need to handle the free here.

```c
                                zio_nowait(zio_vdev_child_io(zio, zio->io_bp,
                                    mc->mc_vd, mc->mc_offset,
                                    abd_alloc_sametype(zio->io_abd,
                                    zio->io_size), zio->io_size,
                                    zio->io_type, zio->io_priority, 0,
                                    zio->io_bp ? vdev_mirror_scrub_done :
                                    vdev_mirror_child_done, mc))
```

Reply:
OK, I see that it is necessary to clean up these abd structs... but it seems a bit odd to be doing it via the mc_abd in this context. I guess you can't add an `abd_free()` call to `vdev_mirror_child_done()` because you need access to the abd after that is called (i.e. here).